src/GafferImageUI/ImageView.cpp

```diff
@@ -316,34 +1191,130 @@ class ImageView::ColorInspector : public boost::signals::trackable
+			colorInspectorsPlug()->childAddedSignal().connect( boost::bind( &ColorInspector::colorInspectorAdded, this, ::_2 ) );
+			colorInspectorsPlug()->childRemovedSignal().connect( boost::bind( &ColorInspector::colorInspectorRemoved, this, ::_2 ) );
+
+			colorInspectorsPlug()->getChild<ColorInspectorPlug>( 0 )->modePlug()->setValue( 2 );
+
+			view->plugSetSignal().connect( boost::bind( &ColorInspector::plugSet, this, ::_1 ) );
 		}
 
 	private :
 
-		Plug *plug()
+		Gaffer::ArrayPlug *colorInspectorsPlug()
 		{
-			return m_view->getChild<Plug>( "colorInspector" );
+			return m_view->getChild<Plug>( "colorInspector" )->getChild<ArrayPlug>( "inspectors" );
+		}
+
+		const Gaffer::ArrayPlug *colorInspectorsPlug() const
+		{
+			return m_view->getChild<Plug>( "colorInspector" )->getChild<ArrayPlug>( "inspectors" );
+		}
+
+		void plugSet( Gaffer::Plug *plug )
+		{
+			if( plug->parent()->parent() == colorInspectorsPlug() )
+			{
+				ColorInspectorPlug *colorInspector = IECore::runTimeCast<ColorInspectorPlug>( plug->parent() );
+				if( plug == colorInspector->modePlug() )
+				{
+					if( colorInspector->modePlug()->getValue() == 0 )
+					{
+						colorInspector->pixelPlug()->setValue( colorInspector->regionPlug()->getValue().center() );
+					}
+					else if( colorInspector->modePlug()->getValue() == 1 )
+					{
+						V2i pixel = colorInspector->pixelPlug()->getValue();
+						colorInspector->regionPlug()->setValue( Box2i( pixel - V2i( 50 ), pixel + V2i( 50 ) ) );
+					}
+					colorInspectorRemoved( colorInspector );
+					colorInspectorAdded( colorInspector );
+				}
+			}
+		}
+
+		void colorInspectorAdded( GraphComponent *colorInspector )
+		{
+			ColorInspectorPlug *colorInspectorTyped = IECore::runTimeCast<ColorInspectorPlug>( colorInspector );
+			if( colorInspectorTyped->modePlug()->getValue() == 0 )
+			{
+				V2iGadget::Ptr r = new V2iGadget( colorInspectorTyped->pixelPlug(), colorInspector->getName().value().substr( 1 ) );
+				r->deleteClickedSignal().connect( boost::bind( &ColorInspector::deleteClicked, this, ::_1 ) );
+				m_view->viewportGadget()->addChild( r );
+			}
+			else
+			{
+				Box2iGadget::Ptr r = new Box2iGadget( IECore::runTimeCast<ColorInspectorPlug>( colorInspector )->regionPlug(), colorInspector->getName().value().substr( 1 ) );
+				r->deleteClickedSignal().connect( boost::bind( &ColorInspector::deleteClicked, this, ::_1 ) );
+				m_view->viewportGadget()->addChild( r );
+			}
+		}
+
+		void colorInspectorRemoved( GraphComponent *colorInspector )
+		{
+			ColorInspectorPlug *colorInspectorTyped = IECore::runTimeCast<ColorInspectorPlug>( colorInspector );
+			for( auto &i : m_view->viewportGadget()->children() )
+			{
+				if( (int)i->typeId() == (int)Box2iGadgetTypeId )
+				{
+					if( IECore::runTimeCast<Box2iGadget>( i.get() )->getPlug() == colorInspectorTyped->regionPlug() )
+					{
+						m_view->viewportGadget()->removeChild( i );
```

Comment:
`removeChild()` invalidates iterators of `children()`, so I think this needs to happen outside the loop.

Reply:
It returns directly after this, so there is no use of the invalidated iterator.

I did think a bit about whether only taking the first one we find could somehow be a problem, but we seem pretty reliant on a one-to-one mapping between ColorInspectorPlugs and the gadgets anyway.

Reply:
Ah, yep, sorry, I missed that.